src/devices/src/virtio/block/device.rs

```diff
@@ -1172,6 +1188,104 @@ pub(crate) mod tests {
         }
     }
 
+    fn add_flush_requests_batch(
+        block: &mut Block,
+        mem: &GuestMemoryMmap,
+        vq: &VirtQueue,
+        count: u16,
+    ) {
+        vq.avail.idx.set(0);
+        vq.used.idx.set(0);
+        set_queue(block, 0, vq.create_queue());
+
+        let hdr_addr = vq
+            .end()
+            .checked_align_up(std::mem::align_of::<RequestHeader>() as u64)
+            .unwrap();
+        // Write request header. All requests will use the same header.
+        mem.write_obj(RequestHeader::new(VIRTIO_BLK_T_FLUSH, 0), hdr_addr)
+            .unwrap();
+
+        let mut status_addr = hdr_addr
+            .checked_add(std::mem::size_of::<RequestHeader>() as u64)
+            .unwrap()
+            .checked_align_up(4)
+            .unwrap();
+
+        for i in 0..count {
+            let idx = i * 2;
+
+            let hdr_desc = &vq.dtable[idx as usize];
+            hdr_desc.addr.set(hdr_addr.0);
+            hdr_desc.flags.set(VIRTQ_DESC_F_NEXT);
+            hdr_desc.next.set(idx + 1);
+
+            let status_desc = &vq.dtable[idx as usize + 1];
+            status_desc.addr.set(status_addr.0);
+            status_desc.flags.set(VIRTQ_DESC_F_WRITE);
+            status_desc.len.set(4);
+            status_addr = status_addr.checked_add(4).unwrap();
+
+            vq.avail.ring[i as usize].set(idx);
+            vq.avail.idx.set(i + 1);
+        }
+    }
+
+    fn check_flush_requests_batch(count: u16, mem: &GuestMemoryMmap, vq: &VirtQueue) {
+        let used_idx = vq.used.idx.get();
+        assert_eq!(used_idx, count);
+
+        for i in 0..count {
+            let used = vq.used.ring[i as usize].get();
+            let status_addr = vq.dtable[used.id as usize + 1].addr.get();
+            assert_eq!(used.len, 1);
+            assert_eq!(
+                mem.read_obj::<u8>(GuestAddress(status_addr)).unwrap(),
+                VIRTIO_BLK_S_OK as u8
+            );
+        }
+    }
+
+    #[test]
+    fn test_io_engine_throttling() {
+        // skip this test if kernel < 5.10 since in this case the sync engine will be used.
+        if is_kernel_lt_5_10() {
+            return;
+        }
+
+        let mut block = default_block();
+
+        let mem = default_mem();
+        let vq = VirtQueue::new(GuestAddress(0), &mem, IO_URING_NUM_ENTRIES * 4);
```

Comment:
why `IO_URING_NUM_ENTRIES * 4` ?

Reply:
Since each flush request spreads across 2 virtio descriptors we need (IO_URING_NUM_ENTRIES +1) * 2 virtio descriptors in order to be able to push IO_URING_NUM_ENTRIES flush requets. But the size of the virtio queue has to be a multiple of 2. Hence IO_URING_NUM_ENTRIES * 4 .